File: php/src/EvoCore.php
<?php
/*
 * Copyright (c) 2026 Talon Contributors
 * Author: dark.lijin@gmail.com
 * Licensed under the Talon Community Dual License Agreement.
 * See the LICENSE file in the project root for full license information.
 */

declare(strict_types=1);

namespace Talon\EvoCore;

/**
 * Talon EvoCore — AI 自进化运行时 PHP SDK。
 *
 * 两行代码，赋予 AI 进化的灵魂。
 *
 * @example
 * $evo = new EvoCore('./data');
 * $result = $evo->learn(['domain' => 'coding', 'task_type' => 'review', 'success' => true]);
 */
class EvoCore
{
    private \FFI $ffi;
    private int $id;

    public function __construct(string $dbPath, ?array $config = null)
    {
        $headerDef = <<<CDEF
        char* evo_open(const char* db_path, const char* config_json);
        char* evo_execute(unsigned long long id, const char* action, const char* params_json);
        char* evo_close(unsigned long long id);
        void  evo_free_string(char* ptr);
        CDEF;

        $libPath = $this->findLibrary();
        $this->ffi = \FFI::cdef($headerDef, $libPath);

        $configJson = $config !== null ? json_encode($config) : null;
        $result = $this->parseResult($this->ffi->evo_open($dbPath, $configJson));
        $this->id = (int) $result['id'];
    }

    /** 执行后学习。 */
    public function learn(array $input): array
    {
        return $this->execute('learn', $input);
    }

    /** 推荐策略。 */
    public function recommendStrategy(array $signals): array
    {
        return $this->execute('recommend_strategy', ['signals' => $signals]);
    }

    /** 技能进化。 */
    public function evolveSkill(string $skillName): ?array
    {
        $result = $this->execute('evolve_skill', ['skill_name' => $skillName]);
        return isset($result['skipped']) ? null : $result;
    }

    /** 个性快照。 */
    public function personalitySnapshot(): array
    {
        return $this->execute('personality_snapshot', []);
    }

    /** 进化报告。 */
    public function evolutionReport(int $slowThresholdMs = 5000): array
    {
        return $this->execute('evolution_report', ['slow_threshold_ms' => $slowThresholdMs]);
    }

    /** 关闭实例。 */
    public function close(): void
    {
        if (isset($this->id)) {
            $this->ffi->evo_close($this->id);
            unset($this->id);
        }
    }

    public function __destruct()
    {
        $this->close();
    }

    private function execute(string $action, array $params): array
    {
        $raw = $this->ffi->evo_execute($this->id, $action, json_encode($params));
        return $this->parseResult($raw);
    }

    private function parseResult(string $raw): array
    {
        $data = json_decode($raw, true);
        if (!($data['ok'] ?? false)) {
            throw new \RuntimeException('[evocore] ' . ($data['error'] ?? 'unknown error'));
        }
        return $data['data'] ?? [];
    }

    private function findLibrary(): string
    {
        $env = getenv('EVOCORE_LIB_PATH');
        if ($env && file_exists($env)) return $env;

        $libName = PHP_OS_FAMILY === 'Darwin' ? 'libevocore.dylib'
            : (PHP_OS_FAMILY === 'Windows' ? 'evocore.dll' : 'libevocore.so');
        $pkgPath = __DIR__ . '/../native/' . $libName;
        if (file_exists($pkgPath)) return realpath($pkgPath);

        throw new \RuntimeException(
            "[evocore] Native library not found. Set EVOCORE_LIB_PATH."
        );
    }
}

File: java/src/main/java/com/talon/evocore/EvoCore.java
/*
 * Copyright (c) 2026 Talon Contributors
 * Author: dark.lijin@gmail.com
 * Licensed under the Talon Community Dual License Agreement.
 * See the LICENSE file in the project root for full license information.
 */
package com.talon.evocore;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.util.Map;

/**
 * Talon EvoCore — AI 自进化运行时 Java SDK。
 *
 * <p>两行代码，赋予 AI 进化的灵魂。</p>
 *
 * <pre>{@code
 * try (EvoCore evo = new EvoCore("./data")) {
 *     JsonObject result = evo.learn(Map.of(
 *         "domain", "coding",
 *         "task_type", "review",
 *         "success", true
 *     ));
 * }
 * }</pre>
 */
public class EvoCore implements AutoCloseable {

    private static final Gson GSON = new Gson();
    private static final EvoCoreNative LIB = EvoCoreNative.INSTANCE;

    private long id;

    /** 创建 EvoCore 实例。 */
    public EvoCore(String dbPath) {
        this(dbPath, null);
    }

    /** 创建 EvoCore 实例（带配置）。 */
    public EvoCore(String dbPath, Object config) {
        String configJson = config != null ? GSON.toJson(config) : null;
        JsonObject result = parseResult(LIB.evo_open(dbPath, configJson));
        this.id = result.get("id").getAsLong();
    }

    /** 执行后学习。 */
    public JsonObject learn(Map<String, Object> input) {
        return execute("learn", input);
    }

    /** 推荐策略。 */
    public JsonObject recommendStrategy(String[] signals) {
        return execute("recommend_strategy", Map.of("signals", signals));
    }

    /** 技能进化。 */
    public JsonObject evolveSkill(String skillName) {
        return execute("evolve_skill", Map.of("skill_name", skillName));
    }

    /** 个性快照。 */
    public JsonObject personalitySnapshot() {
        return execute("personality_snapshot", Map.of());
    }

    /** 进化报告。 */
    public JsonObject evolutionReport(long slowThresholdMs) {
        return execute("evolution_report", Map.of("slow_threshold_ms", slowThresholdMs));
    }

    /** 关闭实例。 */
    @Override
    public void close() {
        if (id != 0) {
            LIB.evo_close(id);
            id = 0;
        }
    }

    private JsonObject execute(String action, Object params) {
        if (id == 0) throw new IllegalStateException("EvoCore instance is closed");
        String paramsJson = GSON.toJson(params);
        return parseResult(LIB.evo_execute(id, action, paramsJson));
    }

    private static JsonObject parseResult(String raw) {
        JsonObject obj = GSON.fromJson(raw, JsonObject.class);
        if (!obj.get("ok").getAsBoolean()) {
            String error = obj.has("error") ? obj.get("error").getAsString() : "unknown error";
            throw new RuntimeException("[evocore] " + error);
        }
        return obj.getAsJsonObject("data");
    }
}
